src/xrpld/app/tx/detail/Credentials.cpp

```
@@ -0,0 +1,366 @@
+                         << (page ? "success" : "failure");
+        if (!page)
+            return tecDIR_FULL;
+        sleCred->setFieldU64(sfIssuerNode, *page);
+
+        adjustOwnerCount(view(), sleIssuer, 1, j_);
+    }
+
+    if (subject == account_)
+    {
+        sleCred->setFieldU32(sfFlags, lsfAccepted);
+    }
+    else
+    {
+        auto const page = view().dirInsert(
+            keylet::ownerDir(subject),
+            credentialKey,
+            describeOwnerDir(subject));
+        JLOG(j_.trace()) << "Adding Credential to owner directory "
+                         << to_string(credentialKey.key) << ": "
+                         << (page ? "success" : "failure");
+        if (!page)
+            return tecDIR_FULL;
+        sleCred->setFieldU64(sfSubjectNode, *page);
+        view().update(view().peek(keylet::account(subject)));
+    }
+
+    view().insert(sleCred);
+
+    return tesSUCCESS;
+}
+
+// ------- DELETE --------------------------
+NotTEC
+CredentialDelete::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureCredentials))
+    {
+        JLOG(ctx.j.trace()) << "featureCredentials is disabled.";
+        return temDISABLED;
+    }
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto const subject = ctx.tx[~sfSubject];
+    auto const issuer = ctx.tx[~sfIssuer];
+
+    if (!subject && !issuer)
+    {
+        // Neither field is present, the transaction is malformed.
+        JLOG(ctx.j.trace()) << "Malformed transaction: "
+                               "No Subject or Issuer fields.";
+        return temMALFORMED;
+    }
+
+    // Make sure that the passed account is valid.
+    if ((subject && subject->isZero()) || (issuer && issuer->isZero()))
+    {
+        JLOG(ctx.j.trace()) << "Malformed transaction: Subject or Issuer "
+                               "field zeroed.";
+        return temINVALID_ACCOUNT_ID;
+    }
+
```

In an effort to fail faster, should the `CredentialType` in `CredentialAccept` and `CredentialDelete` undergo the same validations it does in `CredentialCreate` (namely, length)? cc @gregtatcam

---

They all have different validations. Added credentialType check to `CredentialAccept` and `CredentialDelete`.